### ISSUE TITLE
Irregular job time when performing big queries with Mongoid

### ISSUE THREAD

**Issue by TheNaoX** [mperham/sidekiq#2912]
Hey Guys, I don't know if any of you have had this problem, but, I recently upgraded my Rails application dependencies (including sidekiq) and I started to notice that I'm getting irregular responses from some data processing I'm doing with sidekiq, the reason why I'm reporting this here and not in the mongoid community is that it only happens when I run the queries under the sidekiq environment, I tried running the queries manually and I always get consistent query responses. If you ran into something similar or you have an idea of what could be happening I will really appreciate it! You are the best.

Here are the versions of the gems I'm using.

``` Gemfile
rails (4.2.6)
sidekiq (4.1.1)
mongoid (5.1.2)
```

**Comment by mikegee** [comment]
I think you need to be more specific. I don't know what you mean by "irregular" vs. "consistent" query responses. I suppose you mean responses from Mongo, since you also talk about Mongo, but maybe you mean responses from external APIs? How are the responses problematic? Content or response time?

**Comment by trebor8x** [comment]
Hi there,
yesterday I upgraded a Sidekiq deployment from version 3.5.1 to 4.1.1 and I happen to notice similar effects as TheNaoX does.
Sidekiq reports failures in Mongoid `find` operations. When the job is retried  it sometimes completes without error.

Upon closer inspection I noticed that some objects are stored in the wrong of two MongoDB-Clusters (running MongoDB 3.2.4) by Sidekiq.
When a Sidekiq job is looking for a particular Mongoid object it might look into the wrong cluster.

To give an example of the wrongly located objects: the `Participation` model contains `store_in client: 'slow'`, thus all participations should be stored in the cluster configured by 'slow'.
When searching via rails console for a particular participation in the correct cluster, none is found:

```
Participation.with(client: :slow).unscoped.find '5704...fd'
=> Mongoid::Errors::DocumentNotFound: ...
```

However, looking into the other cluster succeeds:

```
Participation.with(client: :secondary).unscoped.find '5704...fd'
=> #<Participation _id: 5704...fd, panel_id: BSON::ObjectId('...'), ...
```

This might be related to how Mongoid stores options in the Thread.
Compare e.g. https://github.com/mongodb/mongoid/blob/7199e4eee458b8b6777babcb02fc0f01b501d50c/lib/mongoid/clients/options.rb#L91

For the moment I downgraded to Sidekiq 3.5.4
The setup contains the following gems:

```
ruby (2.3.0)
rails (4.2.6)
mongoid (5.1.1)
mongoid-multitenancy (1.1.3)
sidekiq-middleware (0.3.0)
```

**Comment by mperham** [comment]
@jonhyman Do you know of any mongoid issues?

**Comment by TheNaoX** [comment]
Hey Guys, I just want to say that I worked around the issue, I'm still not 100% sure but it seems to be a Mongodb + Wired tiger issue, it doesn't like concurrency when pulling on big collections. So what I did is to remove threads from sidekiq and increase worker processes. That sort of fixed the issue, however I'll keep investigating on how to bring concurrency back.

Thanks a lot for your help!

**Comment by jonhyman** [comment]
I do not, but I run my own mongo-ruby-driver that I wrote which hacks in global connection pooling instead of one connection per "client" (of which `slow` and `secondary` would be two clients above). That may be isolating me from problems, not sure. We're also on my own mongoid fork... reasons being that I optimize for our use case and they won't merge my commits into their code bases.

**Comment by ryansch** [comment]
@jonhyman Do you have it published?  I'd be interested in looking at your changes.

**Comment by jonhyman** [comment]
https://github.com/mongodb/mongo-ruby-driver/compare/master...jonhyman:feature/runner-dead-unauthorized-220?expand=1 is what we're on right now. 

I've had some people email me saying that they can't use the official driver, found mine, and it works great for them.

**Comment by ryansch** [comment]
Thanks!

**Comment by estolfo** [comment]
Hi everyone
I maintain Mongoid and I'd like to help get to the bottom of the issues people are seeing with Sidekiq and Mongoid. Someone people have reported that upgrading the ruby driver (mongo gem) to version 2.2.4 has resolved the issue.
If anyone would like to comment on this [jira ticket](https://jira.mongodb.org/browse/RUBY-1100) with their error, mongoid config, and worker code sample, that would help me a lot.

If I understand correctly, @trebor8x you started seeing issues when you upgraded Sidekiq from 3.5.1 to 4.1.1. Did you upgrade your version of Mongoid as well? If so, what version did you upgrade from and what is your current mongo gem version?

**Comment by trebor8x** [comment]
Thanks for looking into this @estolfo .
The mongoid gem (5.1.1) was not upgraded. The used mongo gem version is 2.2.3.

However, neither do we use the WiredTiger engine nor do I recall observing `not authorized on [db] to execute command` errors.

It seems our problem is tied to the use of `Mongoid.override_client`, which configures the client for all following Mongoid operations. Apparently Sidekiq 4 reuses the thread when performing the next job. Thus, following jobs might use the wrong client configured by `Mongoid.override_client`.
Replacing `override_client` with explicit client configuration in each query seems to fix our problem in the staging environment.

**Comment by estolfo** [comment]
Hi @trebor8x 

Thanks for your response. We've just solved issue in the ruby driver that is likely to also be the cause of your error. Would you mind trying out [this pull request](https://github.com/mongodb/mongo-ruby-driver/pull/759) in your testing environment and let me know if you see any find failures reported by sidekiq? Thanks

**Comment by estolfo** [comment]
Hi @trebor8x, @TheNaoX and anyone looking at this issue,

We have worked with Compose to fix an issue with request routing via proxies in front of users' mongodb clusters that manifested as the problem described in this ticket (queries randomly failing).

I don't see Compose mentioned anywhere in this thread, but if anyone was using Compose and saw this issue, you shouldn't see it anymore as they have just deployed the fix to all mongodb deployments.
Thanks,
Emily

**Comment by TheNaoX** [comment]
@estolfo I'll try today to bring concurrency back to sidekiq and see if it works. Thanks!

**Comment by estolfo** [comment]
Hey @TheNaoX 
Can you confirm that you were using Compose? Thanks!

**Comment by TheNaoX** [comment]
Ohh sorry @estolfo yes, I'm using compose on my production app.

**Comment by estolfo** [comment]
Ok, you'll probably see that the issues have gone away with Compose's recent update to mongodb deployments. Please let me know, thanks!

**Comment by phstc** [comment]
@estolfo I'm not using Sidekiq. I'm using Puma and Shoryuken (similar to Sidekiq) and recently I started to see the same problem for both Puma and Shoryuken. I've updated the JIRA issue https://jira.mongodb.org/browse/RUBY-1100?focusedCommentId=2032175&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2032175.

* MongoDB via Compose.io
* mongo (2.6.2)
* mongoid (6.4.2)
* Heroku
* Puma with concurrency (threads)
* Shoryuken with concurrency (threads)

If I restart the Heroku dynos, my processes start to work again. I haven't changed any configuration related to MongoDB in a very long while, so I strongly suspect it is something outside (Compose/Heroku/Network).
